sintr_analysis_server_example/lib/completion_metrics.dart

```diff
@@ -0,0 +1,228 @@
+library sintr_worker_lib.completion;
+
+import 'dart:math' as math;
+
+// TODO(danrubel) ???
+/// Return the Dart SDK version info for the session being extracted
+String get extractedSdkVersion => 'unknown';
+
+/// Process a log entry and return a string representing the result
+/// or `null` if no result from the given log entry.
+final completionExtraction = (String logEntryText) {
+  if (logEntryText == null || logEntryText == "") return null;
+  if (!logEntryText.startsWith("~")) return null;
+
+  int index1 = logEntryText.indexOf(':', 1);
+  if (index1 == -1) return null;
+  int time = int.parse(logEntryText.substring(1, index1));
+
+  int index2 = logEntryText.indexOf(':', index1 + 1);
+  if (index2 == -1) return null;
+  String msgType = logEntryText.substring(index1 + 1, index2);
+
+  return _processLogMessage(time, msgType, logEntryText.substring(index2 + 1));
+};
+
+/// Reporting any partial completion information
+final completionExtractionFinished = () {
+  var results = [];
+  for (_Completion completion in _requestMap.values) {
+    results.add(_composeExtractionResult(completion, -1));
+  }
+  _requestMap.clear();
+  for (_Completion completion in _notificationMap.values) {
+    results.add(_composeExtractionResult(completion, -2));
+  }
+  _notificationMap.clear();
+  return results;
+};
+
+/// Process log line that was encoded by [_composeExtractionResult]
+/// and return a map of current results
+final completionReduction = (String logLine, Map results) {
```

Comment:
rename to completionReducer?

Consider s/logLine/extractionResult? Otherwise it implies that it's the original log entry, which would be a problematic pattern in map-reduce.